### PR TITLE
MONGOID-5388 - Add Index wildcard_projection option and test for wildcard index

### PR DESCRIPTION
**Pull request by johnnyshields**
Please merge this before https://github.com/mongodb/mongoid/pull/5267

@p-mongo for this one, the documentation has this as `wildcardProjection` (camelCase) but in Mongoid we're using  `wildcard_projection` (snake_case). I guess there is some magick which converts these keys to camelCase somewhere? `partial_filter_expression` for example seems to magically work.

### CODE REVIEW

**Comment by johnnyshields** [comment]
Please run Evergreen?

File: spec/mongoid/indexable_spec.rb [hunk 1]
```diff
@@ -576,6 +579,25 @@ def self.hereditary?
       end
     end
 
+    context "when using a wildcard index" do
+
+      before do
+        klass.index({ '$**': 1 }, wildcard_projection: { _id: 1, username: 0 })
```
Comment (p-mongo) [inline spec/mongoid/indexable_spec.rb:585]: What is preventing this line from existing in the model class?

File: lib/mongoid/indexable/validators/options.rb [hunk 2]
```diff
@@ -28,7 +28,8 @@ module Options
           :text_version,
           :version,
           :partial_filter_expression,
-          :collation
+          :collation,
+          :wildcard_projection
```
Comment (p-mongo) [inline lib/mongoid/indexable/validators/options.rb:32]:
```suggestion
          :wildcard_projection,
```

**Comment by p-mongo** [comment]
Hi @johnnyshields , thank you for the PR.

Can you please add a  test that verifies that the option is sent correctly as wildcardProjection. You can use EventSubscriber for this, here is an example of how to use it: https://github.com/mongodb/mongoid/blob/master/spec/mongoid/contextual/mongo_spec.rb#L909

**Comment by johnnyshields** [comment]
Found the magic--turns out a PR is needed to mongo ruby driver to camelize the name. Will raise a PR there.

**Comment by johnnyshields** [comment]
@p-mongo a change in Ruby Driver is needed to support the conversion of `wildcard_projection` to `wildcardProjection`. I have raised a PR for Ruby Driver here which **includes a test for that conversion**: https://github.com/mongodb/mongo-ruby-driver/pull/2541

I would propose to merge this PR as-is, and document the caveat in the release notes that "driver version XXX is required to use the alias".

The specs in this PR match what is in existence today for partial_filter_expression. It is worth nothing that Mongoid contains:
- no tests for end-to-end index creation for all options.
- no tests for index option aliases, such as `partial_filter_expression` to `partialFilterExpression`.

These are covered in the Ruby Driver tests, so Mongoid currently assumes the Ruby Driver works. If we wish to add such specs to Mongoid, I propose to do so in a follow-up ticket which covers *all indexes*.

**Comment by johnnyshields** [comment]
@p-mongo you've marked this ticket as `oleg-responded`, however, my comments are still awaiting your response.

**Comment by johnnyshields** [comment]
This PR is ready to merge:
- Ruby driver change is merged
- Tests are at coverage-parity with existing related features (`partial_filter_expression`)
- [MONGOID-5397](https://jira.mongodb.org/browse/MONGOID-5397) raised to cover future improvement of index tests, i.e. not specific to this PR.

**Review by p-mongo — CHANGES_REQUESTED** [review]

**Comment by p-mongo** [comment]
This PR requires driver 2.18 to work, thus will go into Mongoid 8.0.

File: spec/mongoid/indexable/specification_spec.rb [hunk 3]
```diff
@@ -98,7 +101,11 @@
     end
 
     it "normalizes the options" do
-      expect(spec.options).to eq(background: true, unique: true)
+      expect(spec.options).to eq(background: true,
+                                 unique: true,
+                                 wildcard_projection: {
+                                   views: 1
+                                 })
```
[the remark this answers is not in the document]
Comment (johnnyshields) [reply spec/mongoid/indexable/specification_spec.rb]: I've removed this, it wasn't needed.

**Review by p-mongo — APPROVED** [review]

**Review by Neilshweky — APPROVED** [review]
LGTM

**Pull request merged**